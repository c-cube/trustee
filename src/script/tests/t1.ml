
let top = parse_top_str ~filename:"t1"
{|fn f(x,y) {
   f(x,y) }
|};;

Format.printf "parsed:@ %a@." (pp_or_error Ast.pp_top) top;;

let top = parse_top_str ~filename:"t1"
{|fn f(x,y) {
  let z1 = f(g(h(x), y), z1, z2);
  g(z1, z1);
  z1
}

f(1, "yolo");

fn g(x) { x }
|};;

Format.printf "parsed:@ %a@." (pp_or_error Ast.pp_top) top;;

let top = parse_top_str ~filename:"t1"
{|fn f(x,y, z, ) {
  while p(x,z) {
    var y = f(x+1,"foo");
    y = z + 1 + $ \x (y z: foo). x (x y) z $;
    break;

    let myexpr = $ \x. f ${x} ${ g(y) } $;
    if false { continue; }
    else if true { echo("elseif"); break; }
    else if !true && !false || foo { return 42; }
    else {
      let res = f(1) + f(2) +  if true { 1 } else { 2 };
      return res;
    }
    return x;
  }
  z
}

call_some_fun(g(:atom, :atom2), "bang!");
|};;

Format.printf "parsed:@ %a@." (pp_or_error Ast.pp_top) top;;



