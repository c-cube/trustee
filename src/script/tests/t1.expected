parsed:
(Ok [(S_fn ("f", ["x"; "y"],
        { bl_items = [(Bl_eval (E_app ("f", [(E_var "x"); (E_var "y")])))] }
        ))
      ])
parsed:
(Ok [(S_fn ("f", ["x"; "y"],
        { bl_items =
          [(Bl_let ("z1",
              (E_app ("f",
                 [(E_app ("g", [(E_app ("h", [(E_var "x")])); (E_var "y")]));
                   (E_var "z1"); (E_var "z2")]
                 ))
              ));
            (Bl_eval (E_app ("g", [(E_var "z1"); (E_var "z1")])));
            (Bl_eval (E_var "z1"))]
          }
        ));
      (S_eval
         (E_app ("f", [(E_const (C_int 1)); (E_const (C_string "yolo"))])));
      (S_fn ("g", ["x"], { bl_items = [(Bl_eval (E_var "x"))] }))])
parsed:
(Ok [(S_fn ("f", ["x"; "y"; "z"],
        { bl_items =
          [(Bl_while ((E_app ("p", [(E_var "x"); (E_var "z")])),
              { bl_items =
                [(Bl_var ("y",
                    (E_app ("f", [(E_var "x"); (E_const (C_string "foo"))]))
                    ));
                  (Bl_assign ("y",
                     (E_binop (Plus, (E_var "z"), (E_const (C_int 1))))));
                  Bl_break;
                  (Bl_eval
                     E_if {test = (E_var "false");
                       then_ = { bl_items = [Bl_continue] };
                       elseif =
                       [((E_var "true"),
                         { bl_items =
                           [(Bl_eval
                               (E_app ("echo",
                                  [(E_const (C_string "elseif"))])));
                             Bl_break]
                           });
                         ((E_binop (And, (E_unop (Not, (E_var "true"))),
                             (E_unop (Not, (E_var "false"))))),
                          { bl_items = [(Bl_return (E_const (C_int 42)))] })
                         ];
                       else_ =
                       (Some { bl_items =
                               [(Bl_let ("res",
                                   E_if {test = (E_var "true");
                                     then_ =
                                     { bl_items =
                                       [(Bl_eval (E_const (C_int 1)))] };
                                     elseif = [];
                                     else_ =
                                     (Some { bl_items =
                                             [(Bl_eval (E_const (C_int 2)))]
                                             })}
                                   ));
                                 (Bl_return (E_var "res"))]
                               })});
                  (Bl_return (E_var "x"))]
                }
              ));
            (Bl_eval (E_var "z"))]
          }
        ));
      (S_eval
         (E_app ("call_some_fun",
            [(E_app ("g",
                [(E_const (C_atom ":atom")); (E_const (C_atom ":atom2"))]));
              (E_const (C_string "bang!"))]
            )))
      ])
